jvmquake.c

```diff
@@ -97,151 +90,146 @@ alloc_thread(JNIEnv *env) {
 static void JNICALL
 vm_init(jvmtiEnv *jvmti, JNIEnv *env, jthread thread)
 {
-    jvmtiError err;
-
-    fprintf(stdout, "JVM Watchdog setting up\n");
-
-    err = (*jvmti)->RunAgentThread(
-        jvmti, alloc_thread(env), &watchdog, NULL, JVMTI_THREAD_MAX_PRIORITY
-    );
-    if (err != JVMTI_ERROR_NONE)
-        return;
+	// We only need to do this when signal==0 (i.e. OOM)
+	if(opt_signal != 0) return;
+
+	fprintf(stdout, "JVM Watchdog setting up\n");
+
+	jvmtiError err = (*jvmti)->RunAgentThread(
+		jvmti, alloc_thread(env), &killer_thread, NULL, JVMTI_THREAD_MAX_PRIORITY
+	);
+	if (err != JVMTI_ERROR_NONE) {
+		fprintf(stderr, "JVM Watchdog could not execute thread. Exiting.\n");
+		exit(1);
+	}
 }
 
 static void JNICALL
 gcStarted(jvmtiEnv *jvmti)
 {
-    clock_gettime(CLOCK_MONOTONIC, &last_start);
-
-    long running_nanos = (
-        NANOS * (last_start.tv_sec - last_end.tv_sec) +
-        (last_start.tv_nsec - last_end.tv_nsec)
-    );
+	clock_gettime(CLOCK_MONOTONIC, &last_start);
 
-    // Token bucket algorithm
-    if (running_nanos > val) { val = 0; }
-    else { val -= running_nanos; }
+	long running_nanos = (
+		NANOS * (last_start.tv_sec - last_end.tv_sec) +
+		(last_start.tv_nsec - last_end.tv_nsec)
+	) * opt_runtime_weight;
 
-    //fprintf(stderr, "Got a GC Start!\n");
-    //fprintf(stderr, "val %ld\n", val);
+	// Token bucket algorithm
+	if (running_nanos > val) { val = 0; }
+	else { val -= running_nanos; }
 }
 
 static void JNICALL
 gcFinished(jvmtiEnv *jvmti) {
-    jvmtiError err;
-
-    clock_gettime(CLOCK_MONOTONIC, &last_end);
-    long gc_nanos = (
-        NANOS * (last_end.tv_sec - last_start.tv_sec) +
-        (last_end.tv_nsec - last_start.tv_nsec)
-    );
-    //fprintf(stderr, "GC Time: %ld\n", gc_nanos);
-
-    // Token bucket algorithm
-    val += gc_nanos;
-
-    if (val > (NANOS * watchdog_gc_seconds)) {
-        // Trigger kill due to excessive GC
-        fprintf(stderr, "Excessive GC: setting flag and notifying!\n");
-        trigger_watchdog = 1;
-        err = (*jvmti)->RawMonitorEnter(jvmti, lock);
-        err = (*jvmti)->RawMonitorNotify(jvmti, lock);
-        err = (*jvmti)->RawMonitorExit(jvmti, lock);
-        if (err != JNI_OK) {
-            fprintf(stderr, "Error exiting monitor");
-        }
-    }
-
-    //fprintf(stderr, "Got a GC End! %d\n", err);
-    //fprintf(stderr, "Token %ld\n", val);
+	jvmtiError err;
+
+	clock_gettime(CLOCK_MONOTONIC, &last_end);
+
+	// Token bucket algorithm
+	long gc_nanos = (
+		NANOS * (last_end.tv_sec - last_start.tv_sec) +
+		(last_end.tv_nsec - last_start.tv_nsec)
+	);
+	val += gc_nanos;
+
+	if (val > opt_gc_threshold) {
+		if(opt_signal != 0) {
+			fprintf(stderr, "Excessive GC: sending signal (%d)\n", opt_signal);
+			raise(opt_signal);
+			raise(SIGKILL);
```

Comment:
do we want to raise `SIGKILL` after raising the optional signal? What happens in the case where we e.g. do an abort?